File: Tutorials and Problems/P2a/Solutions a/LunchList1c/LunchList1c/app/src/main/java/org/feup/apm/lunchlist1c/MainActivity.java
package org.feup.apm.lunchlist1c;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.RadioGroup;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity implements AdapterView.OnItemClickListener {
  List<Restaurant> rests = new ArrayList<>();
  ArrayAdapter<Restaurant> adapter;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    ActionBar bar = getSupportActionBar();
    if (bar != null) {
      bar.setIcon(R.drawable.rest_icon);
      bar.setDisplayShowHomeEnabled(true);
    }
    setContentView(R.layout.activity_main);

    findViewById(R.id.bt_save).setOnClickListener((View v)->onBtSaveClick());

    ListView list = findViewById(R.id.listview);
    adapter = new ArrayAdapter<>(this, android.R.layout.simple_list_item_1, rests);
    list.setAdapter(adapter);
    list.setOnItemClickListener(this);
  }

  void onBtSaveClick() {
    Restaurant r = new Restaurant();
    r.setName(((EditText)findViewById(R.id.ed_name)).getText().toString());
    r.setAddress(((EditText)findViewById(R.id.ed_address)).getText().toString());
    switch (((RadioGroup)findViewById(R.id.rg_types)).getCheckedRadioButtonId()) {
      case R.id.rb_take:
        r.setType("take");
        break;
      case R.id.rb_sit:
        r.setType("sit");
        break;
      case R.id.rb_delivery:
        r.setType("delivery");
        break;
    }
    adapter.add(r);
  }

  @Override
  public void onItemClick(AdapterView<?> adapterView, View view, int pos, long id) {
    Restaurant r = rests.get(pos);

    ((EditText)findViewById(R.id.ed_name)).setText(r.getName());
    ((EditText)findViewById(R.id.ed_address)).setText(r.getAddress());
    RadioGroup types = findViewById(R.id.rg_types);

    if (r.getType().equals("sit")) {
      types.check(R.id.rb_sit);
    }
    else if (r.getType().equals("take")) {
      types.check(R.id.rb_take);
    }
    else {
      types.check(R.id.rb_delivery);
    }
  }
}

File: Tutorials and Problems/3.1. The Lunch Restaurants App/LunchList1/LunchList1c/app/src/main/java/org/feup/apm/lunchlist1c/MainActivity.kt
package org.feup.apm.lunchlist1c

import android.os.Bundle
import android.widget.*
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity() {
  private val rests = arrayListOf<Restaurant>()
  private val adapter by lazy { ArrayAdapter(this, android.R.layout.simple_list_item_1, rests) }
  private val edName by lazy { findViewById<EditText>(R.id.ed_name) }
  private val edAddress by lazy { findViewById<EditText>(R.id.ed_address) }
  private val rgTypes by lazy { findViewById<RadioGroup>(R.id.rg_types) }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    supportActionBar?.setIcon(R.drawable.rest_icon)
    supportActionBar?.setDisplayShowHomeEnabled(true)
    setContentView(R.layout.activity_main)

    findViewById<Button>(R.id.bt_save).setOnClickListener { _ -> onBtSaveClick() }
    val list = findViewById<ListView>(R.id.listview)
    list.adapter = adapter
    list.setOnItemClickListener { _, _, pos, _ -> onRestItemClick(pos) }
  }

  private fun onBtSaveClick() {
    val r = Restaurant()
    r.name = edName.text.toString()
    r.address = edAddress.text.toString()
    r.type = when(rgTypes.checkedRadioButtonId) {
        R.id.rb_take -> "take"
        R.id.rb_sit -> "sit"
        R.id.rb_delivery -> "delivery"
        else -> ""
    }
    adapter.add(r)
  }

  private fun onRestItemClick(pos: Int) {
    val r = rests[pos]
    edName.setText(r.name)
    edAddress.setText(r.address)
    when (r.type) {
      "sit" -> rgTypes.check(R.id.rb_sit)
      "take" -> rgTypes.check(R.id.rb_take)
      "delivery" -> rgTypes.check(R.id.rb_delivery)
    }
  }
}

/*
 *  The Restaurant class, for storing information for each restaurant
 */

class Restaurant {
  var name = ""
  var address = ""
  var type = ""

  override fun toString(): String {
    return name
  }
}
